keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/optimizer/gradient/Adam.java

```diff
@@ -0,0 +1,194 @@
+package io.improbable.keanu.algorithms.variational.optimizer.gradient;
+
+import io.improbable.keanu.algorithms.Variable;
+import io.improbable.keanu.algorithms.VariableReference;
+import io.improbable.keanu.algorithms.variational.optimizer.FitnessFunction;
+import io.improbable.keanu.algorithms.variational.optimizer.FitnessFunctionGradient;
+import io.improbable.keanu.algorithms.variational.optimizer.OptimizedResult;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.ToString;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Implemented as described in https://arxiv.org/pdf/1412.6980.pdf
+ */
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class Adam implements GradientOptimizationAlgorithm {
+
+    public interface ConvergenceChecker {
+        boolean hasConverged(DoubleTensor[] gradient, DoubleTensor[] theta, DoubleTensor[] thetaNext);
+    }
+
+    public static AdamBuilder builder() {
+        return new AdamBuilder();
+    }
+
+    private final ConvergenceChecker convergenceChecker;
+    private final double alpha;
+    private final double beta1;
+    private final double beta2;
+    private final double epsilon;
+
+    @Override
+    public OptimizedResult optimize(List<? extends Variable> latentVariables,
+                                    FitnessFunction fitnessFunction,
+                                    FitnessFunctionGradient fitnessFunctionGradient) {
+
+        DoubleTensor[] theta = getTheta(latentVariables);
+        DoubleTensor[] thetaNext = getZeros(theta);
+        DoubleTensor[] m = getZeros(theta);
+        DoubleTensor[] v = getZeros(theta);
+
+        int t = 0;
+        boolean converged = false;
+
+        int maxIterations = Integer.MAX_VALUE;
+
+        final Map<VariableReference, DoubleTensor> thetaMap = new HashMap<>();
+        final DoubleTensor[] gradients = new DoubleTensor[theta.length];
+
+        while (!converged && t < maxIterations) {
+            t++;
+
+            updateGradients(theta, thetaMap, gradients, latentVariables, fitnessFunctionGradient);
+
+            final double beta1T = (1 - Math.pow(beta1, t));
+            final double beta2T = (1 - Math.pow(beta2, t));
+            final double b = beta1T / Math.sqrt(beta2T);
+
+            for (int i = 0; i < theta.length; i++) {
+
+                m[i] = m[i].times(beta1).plusInPlace(gradients[i].times(1 - beta1));
+                v[i] = v[i].times(beta2).plusInPlace(gradients[i].pow(2).timesInPlace(1 - beta2));
+
+                thetaNext[i] = theta[i].plus(m[i].times(alpha).divInPlace(v[i].sqrt().timesInPlace(b).plusInPlace(epsilon)));
+            }
+
+            converged = convergenceChecker.hasConverged(gradients, theta, thetaNext);
+
+            final DoubleTensor[] temp = theta;
+            theta = thetaNext;
+            thetaNext = temp;
```

Comment:
We need to track both theta and theta next for the convergence check. We use two buffers here with one always containing the previous theta and the other the new theta being calculated. Instead of coping the next buffer to the current buffer I just switch the pointer to the buffers.

Reply:
Resolved offline. This is necessary.